add possibility to choose whether snapshot base volume or copy

cyril-s opened the pull request:
# Description
Add key `base_volume_copy` for `libvirt_volume` resource which controls
whether to invoke [StorageVolCreateXML](https://libvirt.org/html/libvirt-libvirt-storage.html#virStorageVolCreateXML) (defualt) or [StorageVolCreateXMLFrom](https://libvirt.org/html/libvirt-libvirt-storage.html#virStorageVolCreateXMLFrom).

# Motivation
Currently workflow of creating libvirt volumes from base volumes works good for qcow2 images. New volumes are created with `backingStore` xml element in definition which associates them with base qcow2 copy-on-write image and allows them to grow in size without any limitations.
However for `logical` storage pools, `backingStore` element instructs libvirt to create **lvm snapshot** which has the following implications:
1) The size of a snapshot volume cannot be larger than it's base volume
2) The size of a snapshot cannot be smaller than it's base volume (which is limited by [this check](https://github.com/dmacvicar/terraform-provider-libvirt/blob/39fd414b6bccc46c00bbb6d52278fca159e4fe32/libvirt/resource_libvirt_volume.go#L213)). 

So workflow of creating libvirt volumes from base volumes for logical storage pool is not very useful: 
it's impossible to create volume of arbitrary size from base volume.

To address this problem, this PR adds possibility to invoke [StorageVolCreateXMLFrom](https://libvirt.org/html/libvirt-libvirt-storage.html#virStorageVolCreateXMLFrom) which creates new ordinary logical volume of any size and just copies the base volume to it.

# Notes
## Uselessness for `dir` type pools
For qcow2 images this knob is rather useless. For qcow2 file images regardless of user specified size, the final size will be the same as the backing image's size. This is so because libvirt invokes `qemu-img create` followed by `qemu-img convert` in [StorageVolCreateXMLFrom](https://libvirt.org/html/libvirt-libvirt-storage.html#virStorageVolCreateXMLFrom) (rather than `qemu-img dd` or copy and `qemu-img resize`).
This behavior can be tested like this:
```bash
$ qemu-img create -f qcow2 test.qcow2 1024K
$ qemu-img info test.qcow2
image: test.qcow2
file format: qcow2
virtual size: 1 MiB (1048576 bytes)
disk size: 196 KiB
cluster_size: 65536
Format specific information:
    compat: 1.1
    lazy refcounts: false
    refcount bits: 16
    corrupt: false
$ qemu-img create -f qcow2 test-copy.qcow2 2048K
$ qemu-img info test-copy.qcow2
image: test-copy.qcow2
file format: qcow2
virtual size: 2 MiB (2097152 bytes)
disk size: 196 KiB
cluster_size: 65536
Format specific information:
    compat: 1.1
    lazy refcounts: false
    refcount bits: 16
    corrupt: false
$ qemu-img convert -f qcow2 -O qcow2 test.qcow2 test-copy.qcow2
$ qemu-img info test-copy.qcow2
image: test-copy.qcow2
file format: qcow2
virtual size: 1 MiB (1048576 bytes)
disk size: 196 KiB
cluster_size: 65536
Format specific information:
    compat: 1.1
    lazy refcounts: false
    refcount bits: 16
    corrupt: false
```
or by specifying larger size for copied volume in `TestAccLibvirtVolume_BackingStoreCopyDir` test in this PR.

## Other storage backends
Behavior for storage backends other than `dir` and `logical` was not tested.

MalloZup commented:
thx for PR. I will check this soon. stay tuned. :sun_with_face:

lblasc commented:
Just tested this pull request, works great!
Without `base_volume_copy` I'm unable to get VM root partition in specified size, as LVM COW snapshot cannot be resized.

Here is my working main.tf:
```
provider "libvirt" {
#  uri   = "qemu+ssh://root@builder1/system"
  uri = "qemu:///system"
}

provider "cloudinit" {}

resource "libvirt_volume" "ubuntu_focal_image" {
  name = "ubuntu_focal_image"
  pool = "default"
  source = "/var/lib/libvirt/focal-server-cloudimg-amd64-disk-kvm.raw"
}

resource "libvirt_volume" "disk" {
  name = "u1-root"
  base_volume_id = libvirt_volume.ubuntu_focal_image.id
  base_volume_copy = true
  size = 100 * 1024 * 1024 * 1024
}

resource "tls_private_key" "bootstrap_key" {
  algorithm = "RSA"
}

data "cloudinit_config" "user_data" {
  gzip = false
  base64_encode = false
  part {
    content = <<EOF
#cloud-config
hostname: u1
fqdn: u1.test
manage_etc_hosts: true
users:
  - name: root
    ssh-authorized-keys:
      - ${tls_private_key.bootstrap_key.public_key_openssh}
# only cert auth via ssh (console access can still login)
ssh_pwauth: false
disable_root: false
packages:
  - qemu-guest-agent
growpart:
  mode: auto
  devices: ['/']
# written to /var/log/cloud-init-output.log
final_message: "The system is finally up, after $UPTIME seconds"
EOF
  }

}

# Use CloudInit to add the instance
resource "libvirt_cloudinit_disk" "commoninit" {
  name      = "commoninit.iso"
  user_data = data.cloudinit_config.user_data.rendered
}

resource "libvirt_domain" "u1" {
  name   = "u1"
  memory = 32 * 1024
  vcpu   = 32
  autostart = true

  network_interface {
    wait_for_lease = true
    network_name = "default"
  }

  disk {
    volume_id = libvirt_volume.disk.id
    scsi = "true"
  }

  cloudinit = libvirt_cloudinit_disk.commoninit.id

  console {
    type = "pty"
    target_type = "serial"
    target_port = "0"
  }

  graphics {
    type = "spice"
    listen_type = "address"
    autoport = true
  }

  connection {
    host = self.network_interface.0.addresses.0
    type = "ssh"
    user = "root"
    private_key = tls_private_key.bootstrap_key.private_key_pem
    agent = false
  }

  provisioner "remote-exec" {
    inline = ["cloud-init status --wait"]
  }

  provisioner "salt-masterless" {
    local_pillar_roots = "${path.root}/../../../../salt-pillar"
    local_state_tree   = "${path.root}/../../../../salt-states"
    minion_config_file = "minion.conf"
  }

  provisioner "remote-exec" {
    inline = [
      "rm -rf /srv/pillar",
      "rm -rf /srv/salt",
      "systemctl stop salt-minion",
      "systemctl disable salt-minion"
    ]
  }
}
```

dmacvicar reviewed (commented):
Thanks for the PR.

Something I don't get to understand from this feature is why `virStorageVolCreateXMLFrom` takes the backing store to copy as an extra parameter, when there is a `<backingStore>` element already present in the definition. Can you enlighten me?
Why doesn't `virStorageVolCreateXMLFrom`  simply copy the backing store mentioned in the definition? why does it need a second argument?

pznamensky commented:
Super useful feature for us! :tada:

beddari commented:
@dmacvicar seems like this is just how `virStorageVolCreateXMLFrom` is defined? https://libvirt.org/html/libvirt-libvirt-storage.html#virStorageVolCreateXMLFrom 

This feature is really useful and should be considered. A common use case is where you don't want or need a backing store for the resulting instance, just a standalone volume.

cyril-s commented:
@dmacvicar  I've rebased this PR onto main and refactored it a bit. I also decided to add a resize step after volume creation if `base_volume_copy` is `true`, to improve libvirt's behavior with `qcow2` images. See the updated PR description for a detailed explanation.


> Thanks for the PR.
> 
> Something I don't get to understand from this feature is why `virStorageVolCreateXMLFrom` takes the backing store to copy as an extra parameter, when there is a `<backingStore>` element already present in the definition. Can you enlighten me? Why doesn't `virStorageVolCreateXMLFrom` simply copy the backing store mentioned in the definition? why does it need a second argument?

`<backingStore>` isn't added to the volume definition when `base_volume_copy` is `true`, so `virStorageVolCreateXMLFrom` needs to know which image to copy from. `base_volume_copy` is essentially equivalent to `virsh vol-create-from`.

dmitry-sinina commented:
@dmacvicar we are using this patch for a long time for VMs on LVM storage. Could you merge it to mainstream?

dmacvicar commented:
ℹ️ ℹ️ ℹ️ ℹ️ ℹ️ ℹ️

This contribution is relevant to the legacy version of the provider, which is now in the v0.8 branch of this repository.

Future development is based on a new provider, which is not compatible with this one, nor does share code.

As the new provider solves many issues with the legacy, and to make development in my free time more enjoyable, I have decided to close all PRs for the legacy provider, and to ask to check if the contribution would apply to the new one. This also to encourage trying the new version.

- https://registry.terraform.io/providers/dmacvicar/libvirt/latest

and check the documentation:

- https://github.com/dmacvicar/terraform-provider-libvirt/blob/main/README.md
- https://registry.terraform.io/providers/dmacvicar/libvirt/latest/docs

You are free to reopen the PR for v0.8, which is targetted now to the v0.8 branch. We may also start a discussion if we can assemble a team of maintainters for the legacy branch. My efforts will go into the new provider.

I ask you to check the new provider and re-evaluate this contribution. 🙏
